kartothek/io_components/metapartition.py

```diff
@@ -1600,8 +1602,42 @@ def parse_input_to_metapartition(obj, metadata_version=None):
             metadata_version=metadata_version,
         )
     elif isinstance(obj, MetaPartition):
-        return obj
+        mp = obj
     else:
         raise ValueError("Unexpected type: {}".format(type(obj)))
 
+    # Check whether index columns of `mp` are equal to `assert_secondary_indices` to prevent
+    # metadata errors
+    if assert_partition_level_indices is not False:
+        if len(mp.metapartitions) > 1:  # nested metapartition
+            for _mp in mp.metapartitions:
+                _ensure_valid_indices(
+                    secondary_indices=assert_partition_level_indices,
+                    mp_indices=_mp["indices"],
+                )
+        else:
+            _ensure_valid_indices(
+                secondary_indices=assert_partition_level_indices, mp_indices=mp.indices
+            )
     return mp
+
+
+def _ensure_valid_indices(secondary_indices, mp_indices):
+    if not secondary_indices:
+        if mp_indices:
+            raise ValueError(
+                "Incorrect indices provided for dataset.\n"
+                f"Expected index columns: {secondary_indices}"
+                f"Provided index: {mp_indices}"
+            )
+    else:
+        # If the dataset has `secondary_indices` defined, then these indices will be build later so there is no need to
+        # ensure that they are also defined here (on a partition level).
+        # Hence,  we just check that no new indices are defined on the partition level.
+        secondary_indices = set(secondary_indices)
+        if not secondary_indices.issuperset(mp_indices.keys()):
```

Comment:
If the keyword argument is used, I'd like to forbid the explicit input via the dictionary. I'd like to get rid of the explicit input in the next major release anyhow